### ISSUE TITLE
What IP and port from RemoteAddr is the WhoIs function checking against on Tailscale side?

### ISSUE THREAD

**Issue by kdevan** [tailscale/caddy-tailscale#12]
For this line: https://github.com/tailscale/caddy-tailscale/blob/main/module.go#L185

What is the `RemoteAddr` that's being passed checking against on the Tailscale side of things? If it's the Tailscale IPv4 then what is the port supposed to be in this case? If it's an endpoint in the Endpoints section, then where is the correct port supposed to come from?

I think I'm just not understanding what's supposed to be on the other side of the Caddy proxy. In my case I'm trying to put a public domain on the other side of this, so I have an Endpoint IP (client IP) and no port (80 or 443 rather but not a port from the client). Is there any way to make this work with just an Endpoint/client IP?

[module.go](https://github.com/tailscale/caddy-tailscale/blob/main/module.go#L185)
`info, err := client.WhoIs(r.Context(), r.RemoteAddr)`

[localclient.go](https://github.com/tailscale/tailscale/blob/v1.36.2/client/tailscale/localclient.go#L241-L248) Tailscale 1.36.2
```
// WhoIs returns the owner of the remoteAddr, which must be an IP or IP:port.
func (lc *LocalClient) WhoIs(ctx context.Context, remoteAddr string) (*apitype.WhoIsResponse, error) {
	body, err := lc.get200(ctx, "/localapi/v0/whois?addr="+url.QueryEscape(remoteAddr))
	if err != nil {
		return nil, err
	}
	return decodeJSON[*apitype.WhoIsResponse](body)
}
```

The comment above implies it might work with only an IP but without the port we get this error `400 Bad Request: invalid 'addr' parameter`. With a port it will do the authorization check.

**Comment by bradfitz** [comment]
It accepts either a IP or an IP:port.

If it's just a Tailscale IP (CGNAT range or Tailscale IPv6 ULA range) then you don't need a port.

If the IP is the loopback interface, you need the port. This then maps the identity back through the userspace networking proxy

**Comment by kdevan** [comment]
I believe I tried the Tailscale IP, I just passed it directly instead of `RemoteAddr` just to see if it works and I still got the `400 Bad Request: invalid 'addr' parameter` error. But with the information you've provided I'll do some double checking on my end.

The Tailscale IPv4 from the screenshot below is what I'm trying. And it would be very helpful if you could clarify, are the IP's listed in the Endpoints section here from the loopback interface? If so where does that port number come from? How is it determined?

![Screenshot from 2023-03-07 17-29-15](https://user-images.githubusercontent.com/1412685/223595858-c012551d-bef5-44c5-bd92-fba4a8c1c4f6.png)

**Comment by bradfitz** [comment]
You're going to have to be more specific. What value are you sending? It should look like:

- 100.102.103.104 (some IP in the CGNAT range)
- 100.102.103.104:some-port
- 127.0.0.1:some-port

(or IPv6 equivalent)

**Comment by kdevan** [comment]
What I'm trying to send is the client_ip as described [here](https://github.com/caddyserver/caddy/pull/5104). I would be using a trusted proxy set up as described in this pull request, "I decided that the client_ip should always be set (as long as the RemoteAddr is a valid IP address). So in the logs, it will be identical to the remote_ip unless trusted proxies is configured and the proxy passed the client IP in a trusted header."

I see that my own client IP is listed in the Endpoints section for the machine I set up for myself, so I thought maybe the IP:port values in the Endpoints section might be part of this lookup. However, I'm getting the feeling that this is the wrong use case for the WhoIs lookup or that there might need to be an extra step before this to somehow go from client IP to Tailscale IP?
